ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java

```diff
@@ -248,6 +249,32 @@ public ParseContext transform(ParseContext pctx) throws SemanticException {
     return pctx;
   }
 
+  private static List<List<TableScanOperator>> groupTableScanOperators(List<Entry<String, Long>> sortedTables,
+      ArrayListMultimap<String, TableScanOperator> tableNameToOps, int batchSize) {
+    final List<List<TableScanOperator>> batches = new ArrayList<>();
+    for (Entry<String, Long> tablePair : sortedTables) {
+      final String tableName = tablePair.getKey();
+      final List<TableScanOperator> scans = tableNameToOps.get(tableName);
+      if (batchSize == -1) {
+        batches.add(scans);
+        continue;
+      }
+
+      final int limit = scans.size();
+      int from = 0;
+      while (from != limit) {
+        final int to = Math.min(limit, from + batchSize);
+        // We have to copy the list because it is mutated later
+        final List<TableScanOperator> subList = new ArrayList<>(scans.subList(from, to));
+        Preconditions.checkState(!subList.isEmpty());
```

Comment:
Could we change this Preconditions to an assertion or some other expressions that guarantee `!scans.subList(from, to).isEmpty()`? (e.g. `Preconditions.checkArgument(batchSize > 0)` at the top of the method.) Using Preconditions to check the invariant of code, rather than caller's or client's requirements, seems a bit odd to me.

Reply:
The level of intention of this assertion might be lower than what you are thinking. I just added it to ensure I correctly implemented L265-L272 because I am worried too much. I think we can remove the line basically. Will do